queue/abstract/queue_session.lua

```diff
@@ -0,0 +1,192 @@
+local cur_time = util.time()
+    for session_uuid, exp_time in pairs(inactive_sessions) do
+        if cur_time >= exp_time then
+            if queue_session._on_session_remove ~= nil then
+                queue_session._on_session_remove(session_uuid)
+            end
+            inactive_sessions[session_uuid] = nil
+        end
+    end
+end
+
+--- Create an expiration fiber to cleanup expired sessions.
+local function create_expiration_fiber()
+    local exp_fiber = fiber.create(function()
+        while true do
+            if box.info.ro == false then
+                local status, err = pcall(cleanup_inactive_sessions)
+                if status == false then
+                    log.error('An error occured while cleanup the sessions: %s',
+                        err)
+                end
+            end
+            fiber.sleep(1)
+        end
+    end)
+
+    return exp_fiber
+end
+
+--- Identify the session.
+-- If session_uuid == nil - create a new UUID for the connection
+-- and return it, otherwise connect to the existing one.
+local function identify(conn_id, session_uuid)
+    local queue_session_ids = box.space._queue_session_ids
+    local session_ids = queue_session_ids.index.box_sid:get(conn_id)
+    local cur_uuid = session_ids and session_ids[2]
+
+    if session_uuid == nil and cur_uuid == nil then
+        -- Generate new UUID for the session.
+        cur_uuid = uuid.bin()
+        queue_session_ids:insert{conn_id, cur_uuid}
+    elseif session_uuid ~= nil then
+        --Validate UUID.
+        uuid.frombin(session_uuid)
+
+        -- identify using a previously created session.
+        -- Check that a session with this uuid exists.
+        local ids_by_uuid = queue_session_ids.index.uuid:select(
+            session_uuid, { limit = 1 })[1]
+        if ids_by_uuid == nil and inactive_sessions[session_uuid] == nil then
+            error('The UUID ' .. uuid.frombin(session_uuid):str() ..
+                ' is unknown.')
+        end
+
+        if cur_uuid ~= session_uuid then
+            if cur_uuid ~= nil then
+                queue_session.remove_connection(conn_id)
+            end
+            queue_session_ids:insert({conn_id, session_uuid})
+            cur_uuid = session_uuid
+        end
+    end
+
+    -- Exclude the session from inactive.
+    inactive_sessions[cur_uuid] = nil
+
+    return cur_uuid
+end
+
+local function on_session_remove(callback)
+    if type(callback) ~= 'function' then
+        error('The "on_session_remove" argument type must be "function"')
+    end
+    queue_session._on_session_remove = callback
+end
+
+--- Remove a connection from the list of active connections and
+-- release its tasks if necessary.
+local function remove_connection(conn_id)
+    local queue_session_ids = box.space._queue_session_ids
+    local session_uuid = queue_session.get_uuid(conn_id)
+
+    queue_session_ids:delete{conn_id}
+    local session_ids = queue_session_ids.index.uuid:select(session_uuid,
```

Comment:
AFAIK you have some performance test for this feature. Could you please check is it reasonable or not to use "count" instead of "select" here? And don't create Lua objects.

Reply:
No, I was wrong. Event count = 100 it work slower than select.
```lua
local uuid = require('uuid')
local clock = require('clock')

box.cfg{}

local s = box.schema.space.create('test')
s:create_index('pk')
s:create_index('uuid', {unique = false, parts = {2, 'string'}})

box.begin()
local uuid_bin = uuid.bin()
for i = 1, 1e2 do
    s:replace{i, uuid_bin}
end
box.commit()

local s_select = clock.time()
for _ = 1, 1e5 do
    s.index.uuid:select({uuid_bin}, {limit = 1})
end
print('select', clock.time() - s_select)

local s_count = clock.time()
for _ = 1, 1e5 do
    s.index.uuid:count({uuid_bin}, {limit = 1})
end
print('count', clock.time() - s_count)

-- select	0.098401069641113
-- count	0.32897591590881
```